### ISSUE TITLE
allow other effects besides Future, perhaps Task

### ISSUE THREAD

**Issue by aappddeevv** [cb372/scalacache#131]
Task from monix or fs2. This requires abstracting out the Future part. fs2 did this so you do not have to use Task if you do not want to.

**Comment by cb372** [comment]
I've been wanting to do this for ages. Just the other day I was sketching out a new design for ScalaCache that would support this.

Here's [an example](https://github.com/cb372/scalacache2/blob/85b57e2f113e97b9c8a78f71419379e59f26e703/modules/caffeine/src/test/scala/scalacache/Example.scala#L29-L40) of how using the new API would look.

scalacache-core would provide support for `Id` and Scala `Future` out of the box, but it would be easy for people to choose other effects just by implementing a "mode" (like [this one](https://github.com/cb372/scalacache2/blob/85b57e2f113e97b9c8a78f71419379e59f26e703/modules/core/src/main/scala/scalacache/Modes.scala#L10-L13)) that provides an instance of cats `Monad` for their effect.

We could even provide integration modules for popular effects like Scalaz `Task` and Monix `Task`.

Other benefits of the API redesign:

* Better type safety - you specify the type of your values once when you construct the `Cache`, not every time you do a `put` or `get`. So there's less risk of writing a `Foo` and then trying to read it back as a `Bar`.
* Codecs make more sense - similarly, you only specify the codec once, when you construct the `Cache`. So less risk of writing something as binary and then trying to read it as JSON.
* Not trying to shoehorn everything into a `Future`, which makes no sense for in-memory caches such as Guava and Caffeine, and actually causes performance overhead due to unnecessary wrapping.
* Ability to write versions of cache backends that are optimised for a particular monad, e.g. see [SyncCaffeineCache](https://github.com/cb372/scalacache2/blob/85b57e2f113e97b9c8a78f71419379e59f26e703/modules/caffeine/src/main/scala/scalacache/caffeine/CaffeineCache.scala#L24-L33) for an example

Overall I'm a lot happier with this new API than with the existing one. The only downside I can see is that it's quite a large breaking change. I can always claim "we haven't made it to version 1.0 yet, so anything goes!" but I know this would affect quite a lot of existing users. I imagine I would release it with a big minor version jump, as say 0.20, while promising to maintain the existing API for a reasonable amount of time while people migrate. Then I'd like to work towards a 1.0 release asap.

Other things I'd like to have before a v1.0:
* Support for better codecs, e.g. JSON with circe
* A microsite with proper tut-checked documentation
* A cache backend that can be used with ScalaJS

As soon as I have time I'll flesh out my prototype a bit more and turn it into a PR to garner some feedback.

/cc @fommil who might be interested in this.

**Comment by aappddeevv** [comment]
I'm not familiar enough the full scalacache api, I did just hack up a scalajs node-cache version but its untested. However, fs2 uses a typeclass to abstract out the effect and my guess is that this would work for you as well. It would force an implicit on most of your API calls, at least where you return or consume an effect. Having said that, it is fairly transparent. I think cats effects (1.0>???) or fs2 (<0.9.7 because its going to use cats effects) defines the typeclass Async that does this but I think all you really need is Monad (for pure) and MonadError (which has a fail). Maybe you can get away with an Applicative if there is an ApplicativeError.

**Comment by aappddeevv** [comment]
I think your API would look more like:
```scala
def get[F[_], V](key: String)(implicit codec: Codec[V, Repr]): F[Option[V]]
```
than the modes link you provided above. Or maybe, the effect is at the Cache class level:
```scala
trait Cache[+F[_]]
```

**Comment by cb372** [comment]
Fixed by #145
